feat: added config sdk and refactored FQ with bug fixes

KaivalyaMDabhadkar opened the pull request:
## Summary

<!-- Brief description of your changes -->

## Type of Change
- [x] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [x] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Health Monitors
- [ ] Core Services
- [x] Fault Management
- [ ] Documentation/CI
- [ ] Other: ____________

## Testing
- [x] Tests pass locally
- [x] Manual testing completed
- [ ] No breaking changes (or documented)

## Checklist
- [x] Self-review completed
- [ ] Documentation updated (if needed)
- [x] Ready for review

## Tests performed on dev cluster manually:
1. Injected an inforom DCGM  error via gpu health monitor on a node (using `dcgmi test --host nvidia-dcgm.gpu-operator:5555 --inject --gpuid 0 -f 84 -v 0`) through the gpu monitor and verified that the node gets cordoned properly and that the `quarantineHealthEvent` annotation gets applied on the node properly for the GPU 0 entity. Also verified that the node gets through the node drainer and fault remediation pipeline properly and that the node gets rebooted through janitor properly. Verified that the node is uncordoned and the annotation removed after the reboot successfully.

2. Manually uncordoned the node after injecting the inforom GPU error. Checked that FQ removed the `quarantineHealthEvent` annotation from the node successfully.

3. Injected a GPU inforom error for GPU 0 first then on GPU 1. Verified that the annotation gets updated properly with both the entities when it is cordoned. Then inserted a healthy event for GPU 1, and verified that the entity GPU 1 gets removed from the annotation, but that the node still remains cordoned. Then inserted a healthy event for GPU 0, and verified that the annotation gets removed from the node and that the node gets uncordoned.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added environment variable configuration with type validation and default values
  * Implemented TOML configuration file loading
  * Added circuit breaker state persistence via Kubernetes ConfigMaps
  * Introduced event watcher for processing health events from MongoDB
  * Expanded metrics exposure for circuit breaker and reconciliation operations

* **Refactor**
  * Centralized component initialization workflow
  * Migrated state persistence to Kubernetes-native storage
  * Reorganized metrics package with improved naming conventions
  * Streamlined node quarantine and uncordon operations
  * Enhanced Kubernetes client integration throughout

* **Tests**
  * Added comprehensive Kubernetes integration tests using envtest

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lalitadithya commented on configmanager/env.go:
```diff
@@ -0,0 +1,256 @@
+	}
+}
+
+func parseInt[T any](valueStr string) (T, error) {
```
What is the purpose of generics here, won't T always be int? What am I missing?

lalitadithya commented on configmanager/env.go:
```diff
@@ -0,0 +1,256 @@
+func parseValue[T any](valueStr string) (T, error) {
+	var zero T
+
+	// Use type assertion to determine the target type
```
Can we use https://pkg.go.dev/fmt#Sscanf here? would it make this logic simpler?

KaivalyaMDabhadkar replied:
fmt.Sscanf uses reflection and format string parsing and is slower and less explicit than the current strconv function. Strconv also provides stronger type safety, we should still use strconv functions imo

lalitadithya commented on configmanager/env.go:
```diff
@@ -0,0 +1,256 @@
+		return spec.DefaultValue, nil
+	}
+
+	description := spec.Description
```
What is the use case for description? Can't these be simple code comments? Do we have logic that hangs off of the description? If we do, then it will be very brittle

lalitadithya commented on configmanager/Makefile:
The makefile structures have changed with the latest changes, please rebase and updated

lalitadithya commented on fault-quarantine-module/pkg/informer/k8s_client.go:
```diff
@@ -0,0 +1,488 @@
+func (c *FaultQuarantineClient) handleUncordon(
+	node *v1.Node, isUnCordon bool, labels map[string]string, nodename string,
+) {
+	if !isUnCordon {
```
Why do we need this variable? It seems like an error check that should have been caught upstream?

KaivalyaMDabhadkar replied:
Removed it

lalitadithya commented on fault-quarantine-module/pkg/informer/node_informer.go:
```diff
@@ -374,96 +304,69 @@ func (ni *NodeInformer) handleDeleteNode(obj interface{}) {
-	}
+		_, err = ni.clientset.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to update node: %w", err)
```
AFAIK the retry doesn't work if you wrap the error, can you please check?

lalitadithya commented on fault-quarantine-module/pkg/informer/node_informer_test.go:
```diff
@@ -42,17 +40,15 @@ func newNode(name string, labels map[string]string, unschedulable bool) *v1.Node
 }
 
 func TestNewNodeInformer(t *testing.T) {
 	clientset := fake.NewSimpleClientset()
```
Can we move this to make use of envtest?

lalitadithya commented on fault-quarantine-module/pkg/reconciler/reconciler_test.go:
```diff
@@ -136,6 +137,67 @@ func (m *mockEvaluator) GetVersion() string {
+	}
+
+	// If no fake client was provided, create one with initial nodes
+	if fakeClient == nil {
```
When would this not be provided?

lalitadithya commented on fault-quarantine-module/pkg/mongodb/event_watcher.go:
```diff
@@ -0,0 +1,221 @@
+	}
+}
+
+func (w *EventWatcher) SetProcessEventCallback(
```
I see, okay.

KaivalyaMDabhadkar committed:
feat: FQ refactorings with upstream changes

- Added config SDK and refactored FQ with bug fixes
- Added fake change stream client for MongoDB testing
- Added e2e tests
- Removed ini methods in configmanager
- Added missing configmanager to makefile and lint-test
- Merged upstream changes and resolved conflicts

lalitadithya commented on fault-quarantine-module/pkg/informer/k8s_client.go:
```diff
@@ -0,0 +1,490 @@
+}
+
+func (c *FaultQuarantineClient) removeTaints(node *v1.Node, taints []config.Taint, nodename string) bool {
+	if len(taints) == 0 {
```
When would the length of taints be 0? Can we perform upstream validations and fail fast?

lalitadithya commented:
/ok to test 4e6c9027200a0efb9cbe6ce035194486e99ac4c1

lalitadithya commented on fault-quarantine-module/go.mod:
```diff
@@ -5,30 +5,35 @@ go 1.24.0
 	k8s.io/api v0.34.1
 	k8s.io/apimachinery v0.34.1
 	k8s.io/client-go v0.34.1
+	k8s.io/klog v1.0.0
```
@KaivalyaMDabhadkar , we switched to klog v2 recently, please make sure this is addressed

KaivalyaMDabhadkar replied:
Updated

KaivalyaMDabhadkar committed:
fix: fixed merge conflicts

KaivalyaMDabhadkar committed:
fix: addressed some reviews

KaivalyaMDabhadkar committed:
Merge remote-tracking branch 'upstream/main' into kdabhadkar/FQ-refactorings

KaivalyaMDabhadkar committed:
fix: fixed merge conflicts

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/informer/node_informer.go:
```diff
@@ -147,61 +135,69 @@ func (ni *NodeInformer) HasSynced() bool {
+	total := ni.totalNodes
+	ni.mutex.RUnlock()
+
+	// Use indexer to efficiently get quarantined nodes
```
Removed it

KaivalyaMDabhadkar committed:
fix: addressed reviews

KaivalyaMDabhadkar committed:
Merge remote-tracking branch 'upstream/main' into kdabhadkar/FQ-refactorings

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/mongodb/event_watcher.go:
```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		return fmt.Errorf("failed to create change stream watcher: %w", err)
+	}
+	defer watcher.Close(ctx)
```
Done

KaivalyaMDabhadkar committed:
fix: addressed reviews

Signed-off-by: Kaivalya Dabhadkar <kdabhadkar@nvidia.com>

KaivalyaMDabhadkar commented on commons/pkg/configmanager/env_test.go:
```diff
@@ -0,0 +1,201 @@
+
+func TestMain(m *testing.M) {
+	os.Exit(m.Run())
+}
```
Done

KaivalyaMDabhadkar commented on commons/pkg/configmanager/loader_test.go:
```diff
@@ -0,0 +1,61 @@
+	err := LoadTOMLConfig(configPath, &cfg)
+	if err != nil {
+		t.Fatalf("failed to load TOML config: %v", err)
+	}
```
Done

KaivalyaMDabhadkar commented on store-client-sdk/pkg/storewatcher/watchStoreMock.go:
```diff
@@ -0,0 +1,170 @@
+
+	if fn != nil {
+		fn(ctx)
+	}
```
Done

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/informer/k8s_client.go:
```diff
@@ -0,0 +1,483 @@
+				uncordonReason = uncordonReason[:55]
+			}
+
+			node.Labels[c.uncordonedReasonLabelKey] = uncordonReason + "-removed"
```
Not necessary

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/reconciler/reconciler_e2e_test.go:
```diff
@@ -0,0 +1,3685 @@
+				}
+			}
+		}
+	}()
```
Done

KaivalyaMDabhadkar committed:
fix: merged conflicts and addressed reviews

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/informer/node_informer.go:
```diff
@@ -15,17 +15,14 @@
+	"context"
 	"fmt"
 	"log/slog"
 	"reflect"
```
Done

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/breaker/breaker.go:
```diff
@@ -406,13 +408,11 @@ func (b *slidingWindowBreaker) calculateBackoffDelay(attempt int,
-	// Get the actual node count from the last attempt to provide accurate error context
 	ctx := context.Background()
-	actualNodes, err := b.cfg.GetTotalNodes(ctx)
+	actualNodes, err := b.cfg.K8sClient.GetTotalNodes(ctx)
```
Done

KaivalyaMDabhadkar committed:
fix: addressed reviews

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/reconciler/reconciler.go:
```diff
@@ -201,617 +214,571 @@ func (r *Reconciler) Start(ctx context.Context) error {
+	newPriority := rulesetsConfig.RuleSetPriorityMap[evalName]
+
+	r.taintUpdateMu.Lock()
+	defer r.taintUpdateMu.Unlock()
```
The mutex correctly protects all accesses to the shared mutable state (taintAppliedMap and taintEffectPriorityMap). The lock is acquired before reading these maps

KaivalyaMDabhadkar committed:
fix: addressed reviews

KaivalyaMDabhadkar committed:
Merge remote-tracking branch 'upstream/main' into kdabhadkar/FQ-refactorings

lalitadithya commented on fault-quarantine-module/pkg/informer/k8s_client.go:
```diff
@@ -0,0 +1,462 @@
+			return err
+		}
+
+		slog.Debug("Updated node (eventual consistency)", "node", nodeName)
```
do we need to specify `(eventual consistency)`?

KaivalyaMDabhadkar committed:
fix: removed redundant code comments

lalitadithya commented on fault-quarantine-module/pkg/reconciler/reconciler_e2e_test.go:
```diff
@@ -0,0 +1,3698 @@
+			return false
+		}
+		return healthEventsMap.Count() == 2
+	}, 10*time.Second, 200*time.Millisecond, "Should track 2 GPUs")
```
Can we make 10*time.Second, 200*time.Millisecond consts and use them everywhere?

KaivalyaMDabhadkar replied:
Done

KaivalyaMDabhadkar committed:
fix: added better error handling in reconciler

lalitadithya commented on fault-quarantine-module/pkg/reconciler/reconciler_e2e_test.go:
```diff
@@ -0,0 +1,3698 @@
+
+	_, mockWatcher, getStatus, _ := setupE2EReconciler(t, ctx, tomlConfig, nil)
+
+	// Start concurrent updates that will create conflicts during unquarantine
```
Same comment as earlier

KaivalyaMDabhadkar committed:
fix: resolved conflicts

lalitadithya commented on store-client-sdk/pkg/storewatcher/watchStoreMock.go:
please use `_` we renamed the files to follow this pattern recently

mchmarny committed:
Merge branch 'main' into kdabhadkar/FQ-refactorings

KaivalyaMDabhadkar committed:
Merge remote-tracking branch 'upstream/main' into kdabhadkar/FQ-refactorings

KaivalyaMDabhadkar commented on commons/pkg/configmanager/env.go:
```diff
@@ -0,0 +1,243 @@
+	}
+}
+
+func parseAndConvert[T any](value any, err error) (T, error) {
```
It is created to take care of the cyclomatic linter errors, if we inline the logic then the linter throws an error

KaivalyaMDabhadkar commented on fault-quarantine-module/main.go:
```diff
@@ -54,216 +49,53 @@ func main() {
-	namespace, mongoURI, mongoDatabase, mongoCollection, tokenDatabase, tokenCollection, err := loadEnvConfig()
-	if err != nil {
-		return err
+	if _, err := configmanager.GetEnvVar[string]("POD_NAMESPACE"); err != nil {
```
Remnant of conflict merge, removed it

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/breaker/breaker.go:
```diff
@@ -357,15 +362,11 @@ func (b *slidingWindowBreaker) handleSuccessfulNodeCount(totalNodes, attempt int
-func (b *slidingWindowBreaker) handleZeroNodes(attempt, maxRetries int) error {
-	lastErr := fmt.Errorf("GetTotalNodes returned 0 nodes (likely NodeInformer cache not synced yet)")
-
+func (b *slidingWindowBreaker) handleZeroNodes(attempt, maxRetries int) {
```
Inlined it

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/reconciler/reconciler.go:
```diff
@@ -201,617 +214,571 @@ func (r *Reconciler) Start(ctx context.Context) error {
+}
+
+func (r *Reconciler) hasExistingQuarantine(nodeName string) (map[string]string, bool) {
+	annotations, annErr := r.getNodeQuarantineAnnotations(nodeName)
```
Done

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/informer/k8s_client_interface.go:
```diff
@@ -12,28 +12,28 @@
+	UpdateNode(ctx context.Context, nodeName string, updateFn func(*v1.Node) error) error
+	EnsureCircuitBreakerConfigMap(ctx context.Context, name, namespace string, initialStatus breaker.State) error
+	ReadCircuitBreakerState(ctx context.Context, name, namespace string) (breaker.State, error)
+	WriteCircuitBreakerState(ctx context.Context, name, namespace, status breaker.State) error
```
Fixed

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/evaluator/rule_set_evaluator_test.go:
```diff
@@ -289,11 +288,9 @@ func TestInitializeRuleSetEvaluators(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			evaluators, err := InitializeRuleSetEvaluators(tt.ruleSets, clientset, nil)
+			evaluators, err := InitializeRuleSetEvaluators(tt.ruleSets, nil)
```
All callers are using this currently, nothing is breaking

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/evaluator/rule_set_evaluator_test.go:
```diff
@@ -365,10 +362,9 @@ func TestCreateEvaluators(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			evaluators, err := createEvaluators(tt.rules, clientset, nil)
+			evaluators, err := createEvaluators(tt.rules, nil)
```
There's no breaking change

KaivalyaMDabhadkar commented on commons/pkg/configmanager/env_test.go:
```diff
@@ -0,0 +1,393 @@
+
+	if _, exists := results["MISSING_OPTIONAL_EMPTY_DEFAULT"]; exists {
+		t.Error("optional var with empty default should not be in results map")
+	}
```
Added better documentation

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/reconciler/reconciler.go:
```diff
@@ -201,620 +214,557 @@ func (r *Reconciler) Start(ctx context.Context) error {
-		for _, eval := range ruleSetEvals {
-			wg.Add(1)
+		return
+	}
```
intentional

KaivalyaMDabhadkar commented on fault-quarantine-module/pkg/reconciler/reconciler.go:
```diff
@@ -201,620 +214,557 @@ func (r *Reconciler) Start(ctx context.Context) error {
+
+	if err := r.addHealthEventAnnotation(healthEvents, annotationsMap); err != nil {
+		return nil
+	}
```
Not necessary

KaivalyaMDabhadkar committed:
fix: addressed review comments

KaivalyaMDabhadkar committed:
Merge branch 'kdabhadkar/FQ-refactorings' of github.com:KaivalyaMDabhadkar/NVSentinel into kdabhadkar/FQ-refactorings

KaivalyaMDabhadkar committed:
fix: fixed go.mod in different components

KaivalyaMDabhadkar committed:
Merge remote-tracking branch 'upstream/main' into kdabhadkar/FQ-refactorings

KaivalyaMDabhadkar committed:
fix: addressed review comment

KaivalyaMDabhadkar committed:
Merge remote-tracking branch 'upstream/main' into kdabhadkar/FQ-refactorings
